Bottom: 570a5f8843170d4961a2877adf3908223729b001
Top:    495f8d2061fc24f9b991ab45271920cc0164395f
Author: Andrew Cooper <andrew.cooper@novatechweb.com>
Date:   2018-08-06 15:58:56 -0500

Refresh of tls-support

---
diff --git a/docker/docker-openldap/Dockerfile b/docker/docker-openldap/Dockerfile
index 52253cab..a3caa4c0 100644
--- a/docker/docker-openldap/Dockerfile
+++ b/docker/docker-openldap/Dockerfile
@@ -13,6 +13,7 @@ RUN apt-get update && \
     DEBIAN_FRONTEND=noninteractive apt-get install --yes --no-install-recommends \
         ldap-utils=${OPENLDAP_VERSION}* \
         slapd=${OPENLDAP_VERSION}* \
+        ca-certificates  \
     && \
     apt-get clean && \
     rm -rf /var/lib/apt/lists/*
diff --git a/roles/docker-openldap/tasks/main.yml b/roles/docker-openldap/tasks/main.yml
index 98149212..5d2efff5 100644
--- a/roles/docker-openldap/tasks/main.yml
+++ b/roles/docker-openldap/tasks/main.yml
@@ -147,9 +147,12 @@
     hostname: '{{ phpldapadmin_hostname }}'
     ports: '{{ phpldapadmin_port_args }}'
     volumes:
-      - '{{ openssl_dv_name }}:/etc/ssl/private:z'
+      - '{{ openssl_dv_name }}:/container/service/phpldapadmin/assets/apache2/certs:z'
     env:
-      PHPLDAPADMIN_HTTPS: 'false'
+      PHPLDAPADMIN_HTTPS: 'true'
+      PHPLDAPADMIN_HTTPS_CRT_FILENAME: "apache2.crt"
+      PHPLDAPADMIN_HTTPS_KEY_FILENAME: "apache2.key"
+      PHPLDAPADMIN_HTTPS_CA_CRT_FILENAME: "apache2_bundle.crt"
       PHPLDAPADMIN_LDAP_HOSTS: 'ldap'
       LDAP_HOSTNAME: '{{ openldap_hostname }}'
     networks:
@@ -177,3 +180,36 @@
 
 - name: set LDAP proxyagent password
   command: 'docker exec -i {{ openldap_container_name }} /docker-entrypoint.sh set_proxyagent_pass {{ openldap_admin_password | quote }} {{ openldap_proxyagent_password | quote }}'
+
+- name: Configure LDAP TLS 
+  ldap_attr:
+      dn: "cn=config"
+      name: "{{ item.key }}"
+      values: "{{ item.value }}"
+      server_uri: "ldap://{{ openldap_hostname }}:389"
+      bind_dn: cn=admin,dc=novatech
+      bind_pw: "{{ openldap_admin_password }}"
+      state: exact
+  with_dict:
+    olcTLSCACertificateFile: /etc/ssl/private/apache2_bundle.crt
+    olcTLSDHParamFile: /etc/ssl/private/dhparam.pem
+
+- name: set LDAP TLS Certificate Path
+  ldap_attr:
+      dn: "cn=config"
+      name: olcTLSCertificateFile
+      values: /etc/ssl/private/apache2.crt
+      server_uri: "ldap://{{ openldap_hostname }}:389"
+      bind_dn: cn=admin,dc=novatech
+      bind_pw: "{{ openldap_admin_password }}"
+      state: exact
+
+- name: set LDAP TLS Certificate Key
+  ldap_attr:
+      dn: "cn=config"
+      name: olcTLSCertificateKeyFile
+      values: /etc/ssl/private/apache2.key
+      server_uri: "ldap://{{ openldap_hostname }}:389"
+      bind_dn: cn=admin,dc=novatech
+      bind_pw: "{{ openldap_admin_password }}"
+      state: exact
diff --git a/scripts/ldaptest.py b/scripts/ldaptest.py
new file mode 100644
index 00000000..03140575
--- /dev/null
+++ b/scripts/ldaptest.py
@@ -0,0 +1,14 @@
+import ldap
+l = ldap.initialize('ldap://ldap.novatech-llc.com:389',trace_level=2)
+
+l.simple_bind_s(who="cn=admin,dc=novatech",cred="7J1YaZvslJnCmkNXD.IJ")
+
+l.search_s("ou=user,dc=novatech",ldap.SCOPE_SUBTREE,None,["uid"])
+
+l.get_option(ldap.OPT_X_TLS)
+l.get_option(ldap.OPT_X_TLS_ALLOW)
+l.get_option(ldap.OPT_X_TLS_CACERTFILE)
+l.get_option(ldap.OPT_X_TLS_CERTFILE)
+l.get_option(ldap.OPT_X_TLS_KEYFILE)
+
+l.search_s("cn=config",ldap.SCOPE_SUBTREE)
diff --git a/scripts/print-certificate-bundle.sh b/scripts/print-certificate-bundle.sh
new file mode 100644
index 00000000..eb644158
--- /dev/null
+++ b/scripts/print-certificate-bundle.sh
@@ -0,0 +1,8 @@
+#!/bin/sh
+echo
+for cert in "$@"; do
+        echo BEGIN $cert
+        openssl crl2pkcs7 -nocrl -certfile "$cert" | openssl pkcs7 -print_certs -noout
+        echo END $cert
+        echo
+done
diff --git a/tmp/ldap-tls.yml b/tmp/ldap-tls.yml
new file mode 100644
index 00000000..db1d78cb
--- /dev/null
+++ b/tmp/ldap-tls.yml
@@ -0,0 +1,42 @@
+- name: set LDAP TLS Certificate Path
+  ldap_attr:
+      dn: cn=config
+      name: olcTLSCertificateFile
+      values: /etc/ssl/private/exim.crt
+      server_uri: "ldap://{{ openldap_hostname }}:389"
+      bind_dn: cn=admin,dc=novatech
+      state: present
+      bind_pw: "{{ openldap_admin_password }}"
+
+- name: set LDAP TLS Certificate Key
+  ldap_attr:
+      dn: cn=config
+      name: olcTLSCertificateKeyFile
+      values: /etc/ssl/private/exim.key
+      server_uri: "ldap://{{ openldap_hostname }}:389"
+      bind_dn: cn=admin,dc=novatech
+      state: present
+      bind_pw: "{{ openldap_admin_password }}"
+
+
+- name: set LDAP TLS Diffie-Hellman parameters
+  ldap_attr:
+      dn: cn=config 
+      name: olcTLSDHParamFile 
+      values: /etc/ssl/private/dhparam.pem
+      server_uri: "ldap://{{ openldap_hostname }}:389"
+      bind_dn: cn=admin,dc=novatech
+      state: present
+      bind_pw: "{{ openldap_admin_password }}"
+
+
+# Monitoring
+dn: cn=module{0},cn=config
+changetype: modify
+add: olcModuleLoad
+olcModuleLoad: back_monitor
+
+dn: olcDatabase=monitor,cn=config
+objectclass: olcDatabaseConfig
+olcAccess: to * by * read
+olcDatabase: monitor
